File: resources/ios/Sources/TesseractAgent.swift
import Foundation
import UIKit

// Swift port of the Android `TesseractAgent`. Owns everything network: the
// desktop session, the bounded broadcast queue, the HTTP relay transport, and
// command receipt. PHP never talks to the desktop — it only hands envelopes to
// `ingest`; the agent stamps them with identity and a monotonic sequence, then
// ships them off the main thread.
//
// NOT YET COMPILED — authored on a non-mac host. Treat as a reviewed draft:
// the logic mirrors the proven Kotlin agent, but expect a first-build pass.
final class TesseractAgent {

    static let shared = TesseractAgent()

    private init() {}

    private let lock = NSLock()
    private let outboundLock = NSLock()

    private var started = false
    private var connected = false

    private var transport: TesseractDesktopTransport?

    private var outbound: [[String: Any]] = []
    private var commandBuffer: [[String: Any]] = []

    private var seq: Int = 0
    private var sessionId: String?
    private var token: String?
    private let captureId = UUID().uuidString

    private let workerQueue = DispatchQueue(label: "tesseract-agent", qos: .utility)
    private let pumpQueue = DispatchQueue(label: "tesseract-command-pump", qos: .utility)
    private var pumpTimer: DispatchSourceTimer?

    private static let outboundCapacity = 4000
    private static let commandBufferCapacity = 1000
    private static let sendBatchLimit = 50
    private static let sendRetryLimit = 3
    private static let sendRetryDelay = 0.5
    private static let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
    private static let isoFormatterLock = NSLock()

    func connect(config: [String: Any]) {
        lock.lock()
        defer { lock.unlock() }

        if started {
            return
        }
        started = true

        transport = TesseractDesktopTransport(
            host: (config["host"] as? String) ?? "127.0.0.1",
            relayPort: (config["relayPort"] as? Int) ?? 61230,
            appName: (config["appName"] as? String) ?? "Laravel",
            appUrl: (config["appUrl"] as? String) ?? "",
            projectKey: (config["projectKey"] as? String) ?? "",
            projectPath: (config["projectPath"] as? String) ?? "",
            relayUrl: (config["relayUrl"] as? String) ?? "",
            capabilities: (config["capabilities"] as? [Any]) ?? []
        )

        startWorker()
        startCommandPump()

        TesseractCaptureBridge.attach { [weak self] kind, stream, payload in
            self?.emit(kind: kind, stream: stream, payload: payload)
        }
        TesseractDeviceBridge.attach { [weak self] kind, stream, payload in
            self?.emit(kind: kind, stream: stream, payload: payload)
        }
    }

    func ingest(envelopes: [[String: Any]]) -> Bool {
        var count = 0
        for envelope in envelopes {
            enqueue(stamp(envelope))
            count += 1
        }
        return count > 0
    }

    func takeCommands() -> [[String: Any]] {
        outboundLock.lock()
        defer { outboundLock.unlock() }
        let commands = commandBuffer
        commandBuffer.removeAll()
        return commands
    }

    func respond(commandId: String, kind: String?, status: String, detail: [String: Any]?) -> Bool {
        guard let transport, let id = sessionId else {
            return false
        }
        return transport.respondCommand(
            sessionId: id,
            token: token,
            captureId: captureId,
            commandId: commandId,
            kind: kind,
            status: status,
            detail: detail
        )
    }

    func status() -> [String: Any] {
        outboundLock.lock()
        let queued = outbound.count
        outboundLock.unlock()

        return [
            "started": started,
            "connected": connected,
            "sessionId": sessionId ?? NSNull(),
            "queued": queued,
            "transport": transport != nil ? "relay" : "none",
        ]
    }

    func emit(kind: String, stream: String, payload: [String: Any]) {
        var envelope: [String: Any] = [
            "source": "native",
            "stream": stream,
            "kind": kind,
            "payload": payload,
        ]
        envelope = stamp(envelope)
        enqueue(envelope, coalescingKind: kind == "native.view.tree" ? kind : nil)
    }

    /// Stamp identity + ordering at enqueue time. The sessionId is deliberately
    /// NOT set here — envelopes can queue before the session opens; the worker
    /// restamps the live sessionId right before send.
    private func stamp(_ envelope: [String: Any]) -> [String: Any] {
        var stamped = envelope
        outboundLock.lock()
        seq += 1
        stamped["seq"] = seq
        outboundLock.unlock()
        stamped["version"] = 1
        stamped["captureId"] = captureId
        if stamped["sentAt"] == nil {
            stamped["sentAt"] = Self.isoNow()
        }
        return stamped
    }

    private func enqueue(_ envelope: [String: Any], coalescingKind: String? = nil) {
        outboundLock.lock()

        if let coalescingKind,
           let index = outbound.lastIndex(where: { ($0["kind"] as? String) == coalescingKind }) {
            // Replace a queued tree in place: once a newer frame exists the
            // older full snapshot has no live-inspector value and can be large.
            outbound[index] = envelope
            outboundLock.unlock()
            return
        }

        outbound.append(envelope)
        if outbound.count > Self.outboundCapacity {
            outbound.removeFirst(outbound.count - Self.outboundCapacity)
        }
        outboundLock.unlock()
        signalWorker()
    }

    private let workerSemaphore = DispatchSemaphore(value: 0)

    private func signalWorker() {
        workerSemaphore.signal()
    }

    private func ensureSession(maxAttempts: Int) -> Bool {
        if sessionId != nil {
            return true
        }

        var attempt = 0
        while started && sessionId == nil && attempt < maxAttempts {
            attempt += 1
            if let session = transport?.openSession(),
               let id = session["sessionId"] as? String, !id.isEmpty {
                sessionId = id
                token = session["token"] as? String
                connected = true
                NSLog("Tesseract: session opened \(id) capture=\(captureId)")
                return true
            }

            if attempt < maxAttempts {
                Thread.sleep(forTimeInterval: 2.0)
            }
        }

        return sessionId != nil
    }

    private func startWorker() {
        workerQueue.async { [weak self] in
            guard let self else { return }

            // Open the session with retry: at app boot the desktop hub may not
            // be ready for a beat. Envelopes accumulate until the session is
            // live, then drain.
            _ = self.ensureSession(maxAttempts: 30)

            var pendingBatch: [[String: Any]]?
            var pendingFailures = 0

            while self.started {
                if pendingBatch == nil {
                    _ = self.workerSemaphore.wait(timeout: .now() + 2.0)
                }

                // Retry session discovery each pass so a desktop launched
                // after app boot is still picked up.
                if self.sessionId == nil && !self.ensureSession(maxAttempts: 1) {
                    continue
                }

                let batch: [[String: Any]]
                if let pendingBatch {
                    batch = pendingBatch
                } else {
                    self.outboundLock.lock()
                    batch = Array(self.outbound.prefix(Self.sendBatchLimit))
                    self.outbound.removeFirst(min(batch.count, self.outbound.count))
                    self.outboundLock.unlock()
                }

                if batch.isEmpty {
                    continue
                }

                let id = self.sessionId
                let stampedBatch = batch.map { envelope -> [String: Any] in
                    var restamped = envelope
                    restamped["sessionId"] = id ?? ""
                    return restamped
                }

                let ok = self.transport?.sendBatch(
                    sessionId: id,
                    token: self.token,
                    envelopes: stampedBatch
                ) ?? false
                self.connected = ok

                if ok {
                    pendingBatch = nil
                    pendingFailures = 0
                } else {
                    pendingFailures += 1
                    if pendingFailures <= Self.sendRetryLimit {
                        pendingBatch = batch
                        Thread.sleep(forTimeInterval: Self.sendRetryDelay)
                    } else {
                        NSLog("Tesseract: dropping \(batch.count) envelopes after \(pendingFailures) failed sends")
                        pendingBatch = nil
                        pendingFailures = 0
                        self.sessionId = nil
                        self.token = nil
                    }
                }

                self.outboundLock.lock()
                let remaining = self.outbound.count
                self.outboundLock.unlock()
                if remaining > 0 {
                    self.workerSemaphore.signal()
                }
            }
        }
    }

    /// The command pump: polls the desktop for host commands. `native.*`
    /// commands are serviced directly (they drive shell hooks); the rest are
    /// buffered for PHP to drain via `takeCommands`.
    private func startCommandPump() {
        let timer = DispatchSource.makeTimerSource(queue: pumpQueue)
        timer.schedule(deadline: .now() + 1.0, repeating: .milliseconds(300))
        timer.setEventHandler { [weak self] in
            guard let self, self.started, let id = self.sessionId else { return }

            guard let commands = self.transport?.pollCommands(
                sessionId: id,
                token: self.token,
                captureId: self.captureId
            ) else { return }

            for command in commands {
                let kind = (command["kind"] as? String) ?? ""
                if kind.hasPrefix("native.") {
                    let commandId = (command["commandId"] as? String) ?? ""
                    let payload = (command["payload"] as? [String: Any]) ?? [:]
                    let outcome = TesseractCommandBridge.execute(kind: kind, payload: payload)
                    _ = self.respond(
                        commandId: commandId,
                        kind: kind,
                        status: (outcome["ok"] as? Bool) == true ? "ok" : "error",
                        detail: outcome
                    )
                } else {
                    self.outboundLock.lock()
                    self.commandBuffer.append(command)
                    if self.commandBuffer.count > Self.commandBufferCapacity {
                        self.commandBuffer.removeFirst(self.commandBuffer.count - Self.commandBufferCapacity)
                    }
                    self.outboundLock.unlock()
                }
            }
        }
        timer.resume()
        pumpTimer = timer
    }

    private static func isoNow() -> String {
        isoFormatterLock.lock()
        defer { isoFormatterLock.unlock() }

        return isoFormatter.string(from: Date())
    }
}

File: resources/android/src/com/tesseract/collector/TesseractAgent.kt
package com.tesseract.collector

import android.content.Context
import android.util.Log
import com.tesseract.collector.capture.NativeCaptureBridge
import com.tesseract.collector.capture.NativeCommandBridge
import com.tesseract.collector.capture.NativeDeviceBridge
import com.tesseract.collector.transport.DesktopTransport
import com.tesseract.collector.transport.InvalidSessionException
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

/**
 * The native Tesseract agent.
 *
 * Owns everything network: the desktop session, the broadcast queue, the live
 * transport (WS primary, relay fallback), and command receipt. PHP never talks
 * to the desktop — it only hands envelopes to [ingest]; the agent stamps them
 * with session identity and a monotonic sequence, then ships them off the main
 * thread.
 *
 * Lifecycle: [prime] is called from the build-registered init hook so the
 * singleton exists early; [connect] (PHP ignition) supplies the desktop
 * endpoint and starts the worker + capture. Both are idempotent.
 */
object TesseractAgent {

    private val lock = Any()

    @Volatile private var appContext: Context? = null
    @Volatile private var started = false
    @Volatile private var connected = false

    private var transport: DesktopTransport? = null
    private var worker: Thread? = null
    private var pump: Thread? = null
    private var transportConfigSignature: String = ""

    // Bounded so a desktop that never comes up can't grow the queue without
    // limit; overflow drops the oldest envelopes (the newest state wins).
    private const val OUTBOUND_CAPACITY = 4000
    private const val COMMAND_BUFFER_CAPACITY = 1000
    private const val SEND_BATCH_LIMIT = 50

    // A refused batch is held and retried (the desktop hub may be mid-restart)
    // instead of dropped on the first failed send; after the retries are
    // exhausted it is dropped so a dead desktop can't wedge the worker.
    private const val SEND_RETRY_LIMIT = 3
    private const val SEND_RETRY_DELAY_MS = 500L

    private val outbound = LinkedBlockingQueue<JSONObject>(OUTBOUND_CAPACITY)
    private val commandBuffer = LinkedBlockingQueue<JSONObject>(COMMAND_BUFFER_CAPACITY)
    private val seq = AtomicInteger(0)
    private val pendingTreeEnvelope = AtomicReference<JSONObject?>(null)
    private val isoFormatter = object : ThreadLocal<SimpleDateFormat>() {
        override fun initialValue(): SimpleDateFormat {
            return SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
        }
    }

    @Volatile private var sessionId: String? = null
    @Volatile private var token: String? = null
    @Volatile private var captureId: String = UUID.randomUUID().toString()

    fun prime(context: Context) {
        appContext = context
    }

    fun connect(context: Context, config: JSONObject) {
        val nextTransport = DesktopTransport(
            host = config.optString("host", "127.0.0.1"),
            relayPort = config.optInt("relayPort", 61230),
            appName = config.optString("appName", "Laravel"),
            appUrl = config.optString("appUrl", ""),
            projectKey = config.optString("projectKey", ""),
            projectPath = config.optString("projectPath", ""),
            relayUrl = config.optString("relayUrl", ""),
            capabilities = config.optJSONArray("capabilities") ?: JSONArray(),
        )
        val nextSignature = transportSignatureFor(config)

        synchronized(lock) {
            appContext = context
            if (started) {
                if (nextSignature != transportConfigSignature) {
                    transport = nextTransport
                    transportConfigSignature = nextSignature
                    sessionId = null
                    token = null
                    connected = false
                    Log.w("Tesseract", "desktop transport config changed; reopening session")
                }

                return
            }
            started = true
            transport = nextTransport
            transportConfigSignature = nextSignature

            startWorker()
            startCommandPump()

            NativeCaptureBridge.attach(::emit)
            NativeDeviceBridge.attach(context, ::emit)
        }
    }

    private fun transportSignatureFor(config: JSONObject): String {
        return listOf(
            config.optString("host", "127.0.0.1"),
            config.optInt("relayPort", 61230).toString(),
            config.optString("appName", "Laravel"),
            config.optString("appUrl", ""),
            config.optString("projectKey", ""),
            config.optString("projectPath", ""),
            config.optString("relayUrl", ""),
            config.optJSONArray("capabilities")?.toString() ?: "[]",
        ).joinToString("|")
    }

    /**
     * Drain the buffered host -> target commands for PHP to execute and respond
     * to. Returned as a JSON string so nested command payloads survive the
     * bridge intact.
     */
    fun takeCommands(): JSONArray {
        val commands = JSONArray()
        while (true) {
            val command = commandBuffer.poll() ?: break
            commands.put(command)
        }
        return commands
    }

    /** Keep host commands bounded while PHP is unavailable; newest wins. */
    private fun enqueueCommand(command: JSONObject) {
        while (!commandBuffer.offer(command)) {
            if (commandBuffer.poll() == null) {
                commandBuffer.offer(command)
                return
            }
        }
    }

    fun respond(commandId: String, kind: String?, status: String, detail: JSONObject?): Boolean {
        val ok = transport?.respondCommand(sessionId, token, captureId, commandId, kind, status, detail) ?: false
        Log.i("Tesseract", "respond id=$commandId kind=$kind status=$status ok=$ok")
        return ok
    }

    fun ingest(envelopes: JSONArray): Boolean {
        var count = 0
        for (i in 0 until envelopes.length()) {
            val envelope = envelopes.optJSONObject(i) ?: continue
            enqueue(stamp(envelope))
            count++
        }
        return count > 0
    }

    fun emit(kind: String, stream: String, payload: JSONObject) {
        val envelope = JSONObject()
            .put("source", "native")
            .put("stream", stream)
            .put("kind", kind)
            .put("payload", payload)
        stamp(envelope)

        if (kind == "native.view.tree") {
            // A stale tree is useless once a newer frame exists. Keep at most
            // one queued tree so an offline desktop cannot retain thousands of
            // large UI snapshots inside the otherwise count-bounded queue.
            pendingTreeEnvelope.getAndSet(envelope)?.let { outbound.remove(it) }
        }

        enqueue(envelope)
    }

    private fun enqueue(envelope: JSONObject) {
        while (!outbound.offer(envelope)) {
            val evicted = outbound.poll()
            if (evicted == null) {
                // The queue drained between the failed offer and the poll —
                // there's room again, so retry the offer once before giving up.
                outbound.offer(envelope)
                return
            }
            pendingTreeEnvelope.compareAndSet(evicted, null)
        }
    }

    fun status(): JSONObject {
        return JSONObject()
            .put("started", started)
            .put("connected", connected)
            .put("sessionId", sessionId ?: JSONObject.NULL)
            .put("queued", outbound.size)
            .put("transport", transport?.mode() ?: "none")
    }

    private fun resetSession(reason: String) {
        synchronized(lock) {
            sessionId = null
            token = null
            connected = false
        }

        Log.w("Tesseract", "desktop session reset: $reason")
    }

    private fun ensureSession(maxAttempts: Int = 30): Boolean {
        if (sessionId != null) {
            return true
        }

        var attempt = 0

        while (started && sessionId == null && attempt < maxAttempts) {
            attempt++

            try {
                val session = transport?.openSession()
                val id = session?.optString("sessionId") ?: ""

                if (id.isNotEmpty()) {
                    synchronized(lock) {
                        sessionId = id
                        token = session?.optString("token")
                        connected = true
                    }

                    Log.i("Tesseract", "session opened: $id capture=$captureId via ${transport?.mode()}")

                    return true
                }

                Log.w("Tesseract", "session open attempt $attempt got no session; retrying")
                Thread.sleep(2000)
            } catch (t: Throwable) {
                Log.w("Tesseract", "session open attempt $attempt error: ${t.message}")
                Thread.sleep(2000)
            }
        }

        if (sessionId == null) {
            Log.w("Tesseract", "gave up opening a session after $attempt attempts")
        }

        return sessionId != null
    }

    /**
     * Stamp identity + ordering at enqueue time. The sessionId is deliberately
     * NOT set here: envelopes can be queued before the session opens, and a
     * baked-in "" would survive to the wire. The worker restamps the live
     * sessionId right before send.
     */
    private fun stamp(envelope: JSONObject): JSONObject {
        envelope.put("version", 1)
        envelope.put("captureId", captureId)
        envelope.put("seq", seq.incrementAndGet())
        if (!envelope.has("sentAt")) {
            envelope.put("sentAt", isoNow())
        }
        return envelope
    }

    private fun startWorker() {
        val thread = Thread({
            // Open the session with retry: at app boot the desktop hub / adb
            // reverse tunnel may not be ready for a beat. Envelopes accumulate
            // in the queue until the session is live, then drain.
            ensureSession()

            // A batch whose send failed is held here and retried before new
            // envelopes are drained, so a transient desktop hiccup doesn't
            // silently drop it. New envelopes keep queueing meanwhile.
            var pending: MutableList<JSONObject>? = null
            var pendingFailures = 0

            fun holdOrDrop(batch: MutableList<JSONObject>?) {
                pendingFailures++
                if (batch == null || pendingFailures > SEND_RETRY_LIMIT) {
                    if (batch != null) {
                        Log.w(
                            "Tesseract",
                            "dropping batch of ${batch.size} envelopes after $pendingFailures consecutive failed sends",
                        )
                        for (envelope in batch) {
                            pendingTreeEnvelope.compareAndSet(envelope, null)
                        }
                    }
                    pending = null
                    pendingFailures = 0
                } else {
                    pending = batch
                }
            }

            while (started) {
                var batch = pending
                try {
                    if (batch == null) {
                        batch = ArrayList(SEND_BATCH_LIMIT)
                        batch.add(outbound.take())
                        outbound.drainTo(batch, SEND_BATCH_LIMIT - 1)
                    }

                    if (!ensureSession()) {
                        pending = batch
                        Thread.sleep(SEND_RETRY_DELAY_MS)
                        continue
                    }

                    val id = sessionId
                    val envelopes = JSONArray()
                    for (envelope in batch) {
                        envelope.put("sessionId", id ?: "")
                        envelopes.put(envelope)
                    }

                    val ok = transport?.sendBatch(id, token, envelopes) ?: false
                    connected = ok

                    if (ok) {
                        for (envelope in batch) {
                            pendingTreeEnvelope.compareAndSet(envelope, null)
                        }
                        pending = null
                        pendingFailures = 0
                    } else {
                        holdOrDrop(batch)
                        if (pending == null) {
                            for (envelope in batch) {
                                pendingTreeEnvelope.compareAndSet(envelope, null)
                            }
                        }
                        if (pending != null) {
                            Thread.sleep(SEND_RETRY_DELAY_MS)
                        }
                    }
                } catch (ie: InterruptedException) {
                    break
                } catch (stale: InvalidSessionException) {
                    resetSession("desktop rejected telemetry session")
                    pending = batch
                    Thread.sleep(SEND_RETRY_DELAY_MS)
                } catch (t: Throwable) {
                    Log.w("Tesseract", "send failed: ${t.message}")
                    holdOrDrop(batch)
                }
            }
        }, "tesseract-agent")
        thread.isDaemon = true
        worker = thread
        thread.start()
    }

    /**
     * The command pump thread: polls the desktop for host commands and buffers
     * them. Polling off-thread keeps the capture "commandable" on the desktop
     * (each poll marks activity) and decouples fetch from PHP execution. PHP
     * drains the buffer via [takeCommands], executes in its own properly
     * initialized runtime, and answers via [respond].
     *
     * Executing PHP from this thread directly is NOT possible in the current
     * shell: a plugin-created PHPBridge re-runs php_embed_init and crashes the
     * process (TSRM double-init), and the worker runtime is owned by the shell's
     * queue worker. A fully render-independent pump needs the shell to expose a
     * thread-safe plugin execution entry (see install notes).
     */
    private fun startCommandPump() {
        val thread = Thread({
            while (started) {
                try {
                    val id = sessionId
                    if (id == null) {
                        ensureSession(maxAttempts = 1)
                        Thread.sleep(300)
                        continue
                    }

                    val commands = transport?.pollCommands(id, token, captureId)
                    if (commands != null && commands.length() > 0) {
                        for (i in 0 until commands.length()) {
                            commands.optJSONObject(i)?.let { command ->
                                val commandKind = command.optString("kind")
                                if (commandKind.startsWith("native.")) {
                                    // Native commands are serviced by the agent
                                    // (they drive native render bridges), not by PHP.
                                    val commandId = command.optString("commandId")
                                    val payload = command.optJSONObject("payload") ?: JSONObject()
                                    val outcome = when (commandKind) {
                                        "native.highlight" -> NativeCommandBridge.highlight(payload)
                                        "native.clear-highlight" -> NativeCommandBridge.clearHighlight()
                                        "native.scroll-into-view" -> NativeCommandBridge.scrollIntoView(payload)
                                        "native.navigate" -> NativeCommandBridge.navigate(payload)
                                        "native.rotate" -> NativeCommandBridge.rotate(payload)
                                        "native.set-scope" -> NativeCommandBridge.setScope(payload)
                                        "native.set-style" -> NativeCommandBridge.setStyle(payload)
                                        "native.call" -> NativeCommandBridge.call(payload)
                                        else -> NativeCommandBridge.dispatch(payload)
                                    }
                                    respond(
                                        commandId,
                                        commandKind,
                                        if (outcome.optBoolean("ok")) "ok" else "error",
                                        outcome,
                                    )
                                    Log.i("Tesseract", "native command $commandKind -> ${outcome.optString("message")}")
                                } else {
                                    Log.i("Tesseract", "buffered command kind=$commandKind id=${command.optString("commandId")}")
                                    enqueueCommand(command)
                                }
                            }
                        }
                    } else {
                        Thread.sleep(250)
                    }
                } catch (ie: InterruptedException) {
                    break
                } catch (stale: InvalidSessionException) {
                    resetSession("desktop rejected command session")
                    ensureSession(maxAttempts = 3)
                    Thread.sleep(300)
                } catch (t: Throwable) {
                    Log.w("Tesseract", "command pump error: ${t.message}")
                    Thread.sleep(1500)
                }
            }
        }, "tesseract-command-pump")
        thread.isDaemon = true
        pump = thread
        thread.start()
    }

    private fun isoNow(): String {
        return isoFormatter.get()!!.format(Date())
    }
}
